internal/controller/gateway.go

```diff
@@ -31,46 +37,201 @@ import (
+					},
+				},
+			}
+		}
+		r.log.WithField("name", a.GetName()).WithField("namespace", a.GetNamespace()).
+			Info("gateway not owned by contour, bypassing reconciliation.")
+		return []reconcile.Request{}
+	})
+}
+
+// classForGateway returns an nil, error if the gatewayclass referenced by gw does not exist
+// or nil, nil if gw is not owned by Contour. Otherwise, the referenced gatewayclass is returned.
+func (r *gatewayReconciler) classForGateway(gw *gatewayapi_v1alpha1.Gateway) (*gatewayapi_v1alpha1.GatewayClass, error) {
+	gc := &gatewayapi_v1alpha1.GatewayClass{}
+	if err := r.client.Get(r.ctx, types.NamespacedName{Name: gw.Spec.GatewayClassName}, gc); err != nil {
+		return nil, fmt.Errorf("failed to get gatewayclass %s: %w", gw.Spec.GatewayClassName, err)
+	}
+	if !r.isController(gc) {
+		return nil, nil
+	}
+	return gc, nil
+}
+
+// isController returns true if the controller of the provided gc matches Contour's
+// GatewayClass controller string.
+func (r *gatewayReconciler) isController(gc *gatewayapi_v1alpha1.GatewayClass) bool {
+	return gc.Spec.Controller == r.classController
+}
+
 func (r *gatewayReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	r.log.WithField("namespace", request.Namespace).WithField("name", request.Name).Info("reconciling gateway")
 
-	// Fetch the Gateway from the cache.
-	gateway := &gatewayapi_v1alpha1.Gateway{}
-	err := r.client.Get(ctx, request.NamespacedName, gateway)
-	if errors.IsNotFound(err) {
-		r.eventHandler.OnDelete(&gatewayapi_v1alpha1.Gateway{
-			ObjectMeta: metav1.ObjectMeta{
-				Name:      request.Name,
-				Namespace: request.Namespace,
-			},
-		})
+	// Fetch the Gateway.
+	gw := &gatewayapi_v1alpha1.Gateway{}
+	if err := r.client.Get(ctx, request.NamespacedName, gw); err != nil {
+		if errors.IsNotFound(err) {
+			r.log.WithField("name", request.Name).WithField("namespace", request.Namespace).Info("failed to find gateway")
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object, so requeue the request.
+		return reconcile.Result{}, fmt.Errorf("failed to get gateway %s/%s: %w", request.Namespace, request.Name, err)
+	}
+
+	// Check if object is deleted.
+	if !gw.ObjectMeta.DeletionTimestamp.IsZero() {
+		r.eventHandler.OnDelete(gw)
+		if err := r.ensureGatewayDeleted(gw); err != nil {
+			return reconcile.Result{}, fmt.Errorf("failed to ensure deletion of gateway %s/%s: %w",
+				gw.Namespace, gw.Name, err)
+		}
 		return reconcile.Result{}, nil
 	}
 
-	// Pass the new changed object off to the eventHandler.
-	r.eventHandler.OnAdd(gateway)
+	// The gateway is safe to process, so check if it's valid.
+	errs := validation.ValidateGateway(ctx, r.client, gw)
+	if errs != nil {
+		r.log.WithField("name", request.Name).WithField("namespace", request.Namespace).
+			Error("invalid gateway: ", int_errors.ParseFieldErrors(errs))
+	} else {
+		// TODO: Finalize the gateway when managed infrastructure is added.
```

Comment:
We should set `Scheduled` to `status: true` at this point, with a Reason of `Scheduled`, I think? Reference is https://github.com/kubernetes-sigs/gateway-api/blob/16876c7362d58cae211e5599a6ef939ed1b859e9/apis/v1alpha2/gateway_types.go#L574-L614

Reply:
Currently `status.SyncGateway()` sets `Ready=true`, but I'm updating that to `Scheduled=true` since the controller is not provisioning Envoy infra yet. I have also updated `ensureClassFinalizer()` to `addClassFinalizer()` so the func does not update the object. The object's status will now be sync'ed before being updated on the API server.